api/apps/api/src/migrations/api/1645550554581-ImplicitRolesFunctionsAndTriggers.ts

```diff
@@ -0,0 +1,171 @@
+              LEFT OUTER JOIN scenarios s ON G.id = s.id
+              JOIN projects pro ON s.project_id = pro.id
+              JOIN users_projects up ON up.project_id = pro.id AND up.user_id = $1
+          ),
+          R as (
+            SELECT * FROM C EXCEPT SELECT * FROM G
+          )
+        SELECT array((SELECT (id, regexp_replace(role_id, '^project_', 'scenario_'))::role_on_entity FROM GR)) AS grant_on, array((SELECT * FROM R)) AS revoke_on;
+        $$ LANGUAGE 'sql';
+    `);
+
+    await queryRunner.query(`
+    CREATE OR REPLACE FUNCTION apply_scenario_acl_diff_for_user(_user_id uuid, grant_on role_on_entity[], revoke_on uuid[]) RETURNS void as $$
+    DECLARE
+      grant_on_scenario role_on_entity;
+      revoke_on_scenario uuid;
+    BEGIN
+      FOREACH grant_on_scenario IN array grant_on
+      LOOP
+        RAISE NOTICE 'Granting implicit % role on scenario % to user %', grant_on_scenario.role_id, grant_on_scenario.entity_id, _user_id;
+
+        -- first delete, then redo - this is so that we can easily reflect a
+        -- change on user role on the parent project
+        DELETE FROM users_scenarios WHERE user_id = _user_id AND scenario_id = grant_on_scenario.entity_id AND is_implicit IS true;
+        INSERT INTO users_scenarios
+          (user_id, scenario_id, role_id, is_implicit)
+          VALUES
+          (_user_id, grant_on_scenario.entity_id, grant_on_scenario.role_id, true);
+      END LOOP;
+
+      FOREACH revoke_on_scenario IN array revoke_on
+      LOOP
+        RAISE NOTICE 'Revoking implicit role on scenario % from user %', revoke_on_scenario, _user_id;
+
+        DELETE FROM users_scenarios us
+          WHERE us.user_id = _user_id AND scenario_id = revoke_on_scenario;
+      END LOOP;
+    END;
+    $$ LANGUAGE 'plpgsql';
+    `);
+
+    await queryRunner.query(`
+      CREATE OR REPLACE FUNCTION manage_implicit_scenario_roles()
+        RETURNS trigger AS
+      $$
+      DECLARE
+        scenario_id uuid;
+        user_id uuid;
+        is_implicit boolean;
+      BEGIN
+        -- set user_id from row being created or deleted,
+        -- as applicable
+        IF TG_OP = 'INSERT' OR TG_OP = 'UPDATE' THEN
+            user_id := NEW.user_id;
+        ELSIF TG_OP = 'DELETE' THEN
+            user_id := OLD.user_id;
+        END IF;
+
+        PERFORM apply_scenario_acl_diff_for_user(
+          user_id,
+          (select grant_on from compute_implicit_scenario_roles_for_user(user_id)),
+          (select revoke_on from compute_implicit_scenario_roles_for_user(user_id))
+        );
+
+        IF TG_OP = 'INSERT' OR TG_OP = 'UPDATE' THEN
+          RETURN NEW;
+        END IF;
+
+        IF TG_OP = 'DELETE' THEN
+          RETURN OLD;
+        END IF;
+      END;
+      $$ language 'plpgsql';
+    `);
+
+    await queryRunner.query(`
+      CREATE OR REPLACE FUNCTION manage_existing_project_roles_for_scenario()
+      RETURNS trigger as $$
+      DECLARE
+        users_to_grant role_on_entity[];
```

Comment:
mostly to be kind to our future selves - should we create (yet) another type, such as `CREATE TYPE user_and_role AS ( user_id uuid, role_id varchar);` and use this instead? the current implementation works just fine as the shape of data is the same, but whoever may need to read this in a few months may expel fewer WTFs if we reference `user_to_grant.user_id` rather than `user_to_grant.entity_id` in the body of this function.